lib/bolt/module_installer.rb

```diff
@@ -70,130 +78,92 @@ def add(name, modules, puppetfile_path, moduledir, config_path)
+        upgraded.each { |full_name| diff += "#{full_name} #{old[full_name].version} to #{new[full_name].version}\n" }
         @outputter.print_action_step(diff)
       end
 
       # Downgraded modules are those that have an older version in new than old.
-      downgraded = new.select do |title, mod|
-        if old.include?(title)
-          SemanticPuppet::Version.parse(mod.version) < SemanticPuppet::Version.parse(old[title].version)
+      downgraded = new.select do |full_name, mod|
+        if old.include?(full_name)
+          mod.version < old[full_name].version
         end
       end.keys
 
       if downgraded.any?
         diff = "Downgrading the following modules: \n"
-        downgraded.each { |title| diff += "#{title} #{old[title].version} to #{new[title].version}\n" }
+        downgraded.each { |full_name| diff += "#{full_name} #{old[full_name].version} to #{new[full_name].version}\n" }
         @outputter.print_action_step(diff)
       end
 
       # Removed modules are those present in old but not in new.
-      removed = old.reject { |title, _mod| new.include?(title) }.values
+      removed = old.reject { |full_name, _mod| new.include?(full_name) }.values
 
       if removed.any?
         diff = "Removing the following modules:\n"
-        removed.each { |mod| diff += "#{mod.title} #{mod.version}\n" }
+        removed.each { |mod| diff += "#{mod.full_name} #{mod.version}\n" }
         @outputter.print_action_step(diff)
       end
     end
 
     # Installs a project's module dependencies.
     #
-    def install(modules, path, moduledir, force: false, resolve: true)
-      require 'bolt/puppetfile'
-
+    def install(specs, path, moduledir, force: false, resolve: true)
       @outputter.print_message("Installing project modules\n\n")
 
-      puppetfile = Bolt::Puppetfile.new(modules)
-
-      # If the Puppetfile exists, check if it includes specs for each declared
-      # module, erroring if there are any missing. Otherwise, resolve the
-      # module dependencies and write a new Puppetfile. Users can forcibly
-      # overwrite an existing Puppetfile with the '--force' option, or opt to
-      # install the Puppetfile as-is with --no-resolve.
-      #
-      # This is just if resolve is not false (nil should default to true)
       if resolve != false
-        if path.exist? && !force
-          assert_managed_puppetfile(puppetfile, path)
-        else
+        specs = Specs.new(specs)
+
+        # If forcibly installing or if there is no Puppetfile, resolve
+        # and write a Puppetfile.
+        if force || !path.exist?
           @outputter.print_action_step("Resolving module dependencies, this may take a moment")
-          puppetfile.resolve
+          puppetfile = Resolver.new.resolve(specs)
 
           @outputter.print_action_step("Writing Puppetfile at #{path}")
-          # We get here either through 'bolt module install' which uses the
```

Comment:
I think it's useful to leave this here, so that if any of the assumptions this makes changes we have some hint to know we need to change this too.